packages/init/node-async_hooks.js

```diff
@@ -0,0 +1,226 @@
+    const trigger = new Promise(() => {});
+    bootstrapHook.disable();
+
+    // In some versions of Node, async_hooks don't give access to the resource
+    // itself, but to a "wrapper" which is basically hooks metadata for the promise
+    const promisesData = bootstrapData.filter(
+      ({ resource }) => Promise.resolve(resource) === resource,
+    );
+    bootstrapData.length = 0;
+    const { length } = promisesData;
+    if (length > 1) {
+      // console.warn('Found multiple potential candidates');
+    }
+
+    const promiseData = promisesData.find(
+      ({ resource }) => resource === trigger,
+    );
+    if (promiseData) {
+      bootstrapData.push(promiseData);
+    } else if (length) {
+      // console.warn('No candidates matched');
+    }
+  }
+
+  if (bootstrapData.length) {
+    // Normally all promise hooks are disabled in a subsequent microtask
+    // That means Node versions that modify promises at init will still
+    // trigger our proto hooks for promises created in this turn
+    // The following trick will disable the internal promise init hook
+    // However, only do this for destroy modifying versions, since some versions
+    // only modify promises if no destroy hook is requested, and do not correctly
+    // reset the internal init promise hook in those case. (e.g. v14.16.2)
+    const resetHook = createHook({});
+    resetHook.enable();
+    resetHook.disable();
+
+    const { asyncId, triggerAsyncId, resource } = bootstrapData.pop();
+    const symbols = Object.getOwnPropertySymbols(resource);
+    // const { length } = symbols;
+    let found = 0;
+    // if (length !== 3) {
+    //   console.error(`Found ${length} symbols on promise:`, ...symbols);
+    // }
+    symbols.forEach(symbol => {
+      const value = resource[symbol];
+      let type;
+      if (value === asyncId) {
+        type = 'async_id_symbol';
+      } else if (value === triggerAsyncId) {
+        type = 'trigger_async_id_symbol';
+      } else if (typeof value === 'object' && 'destroyed' in value) {
+        type = 'destroyed';
+      } else {
+        // console.error(`Unexpected symbol`, symbol);
+        return;
+      }
+
+      if (setAsyncSymbol(type, symbol)) {
+        found += 1;
+      }
+    });
+    return found;
+  } else {
+    // This node version is not mutating promises
+    return -2;
+  }
+};
+
+const getAsyncHookFallbackState = (promise, create) => {
+  let state = promiseAsyncHookFallbackStates.get(promise);
+  if (!state && create) {
+    state = {
+      [Symbol.nodeAsyncHooksAsyncId]: undefined,
+      [Symbol.nodeAsyncHooksTriggerAsyncId]: undefined,
+    };
+    if (Symbol.nodeAsyncHooksDestroyed) {
+      state[Symbol.nodeAsyncHooksDestroyed] = undefined;
+    }
+    promiseAsyncHookFallbackStates.set(promise, state);
+  }
+  return state;
+};
+
+const setAsyncIdFallback = (promise, symbol, value) => {
+  const state = getAsyncHookFallbackState(promise, true);
```

Comment:
(nitpick) 3 function signatures here have boolean arguments, each meaning different things. using an options object where the boolean is named in the caller as well would limit the lookups one has to make while reading the body of the current function to remember what the boolean did.